### PR TITLE
feat(ext/http): add an op to perform raw HTTP upgrade

### PR DESCRIPTION
**Pull request by mmastrac**
This commit adds new "op_http_upgrade_early", that allows to hijack existing
"Deno.HttpConn" acquired from "Deno.serveHttp" API and performing
a Websocket upgrade on this connection.

This is not a public API and is meant to be used internally in the "ext/node"
polyfills for "http" module.

### CODE REVIEW

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 1]
```diff
@@ -1500,6 +1500,36 @@ declare namespace Deno {
...
 
+  /** **UNSTABLE**: New API, yet to be vetted.
+   *
+   * Allows "hijacking" the connection that the request is associated with. This
+   * can be used to implement protocols that build on top of HTTP (eg.
+   * {@linkcode WebSocket}).
+   *
+   * The returned promise returns the underlying connection.
+   *
+   * ```ts
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1511]: Let's not expose this API on the `Deno` namespace for now. The immediate goal is to wire it up in `ext/node` and I think we don't need to provide public API for now

File: ext/http/01_http.js [hunk 2]
```diff
@@ -471,6 +474,17 @@ function upgradeHttp(req) {
...
 
+function upgradeHttp2(req) {
+  return upgradeHttp2Inner(req);
+}
+
+async function upgradeHttp2Inner(req) {
+  const inner = toInnerRequest(req);
+  const res = await core.opAsync("op_http_upgrade_early", inner[streamRid]);
+  // TODO(mmastrac): We're missing the remote address properties here
+  return new Deno.Conn(res, null, null);
```
Comment (bartlomieju) [inline ext/http/01_http.js:485]: We should address it before landing. I think we can cut corner here - let's pass `conn` parameter to `upgradeHttp2` function and take `remoteAddr` and `localAddr` from it when constructing `Deno.Conn` here.

File: ext/http/01_http.js [hunk 3]
```diff
@@ -471,6 +474,17 @@ function upgradeHttp(req) {
   return req[_deferred].promise;
 }
 
+function upgradeHttp2(req) {
```
Comment (bartlomieju) [inline ext/http/01_http.js:477]: Rename to `upgradeHttpRaw`?

File: ext/http/lib.rs [hunk 4]
```diff
@@ -938,6 +943,197 @@ fn op_http_websocket_accept_header(key: String) -> Result<String, AnyError> {
...
+                .await?
+            }
+            _ => {
+              return Err(http_error("response already started"));
+            }
+          };
+
+          let (rx, tx) = tokio::io::split(upgraded);
+          let rx = Rc::new(AsyncRefCell::new(rx));
+          let tx = Rc::new(AsyncRefCell::new(tx));
```
Comment (bartlomieju) [inline ext/http/lib.rs:1047]: Amazing that it just works with `tokio::io::split`!

File: ext/http/lib.rs [hunk 5]
```diff
@@ -938,6 +943,197 @@ fn op_http_websocket_accept_header(key: String) -> Result<String, AnyError> {
...
+          .await?;
+      }
+    };
+    Ok(())
+  }
+}
+
+impl Resource for EarlyUpgradeSocket {
+  fn name(&self) -> Cow<str> {
+    "earlyUpgradeSocket".into()
```
Comment (bartlomieju) [inline ext/http/lib.rs:1089]: Maybe let's rename it to `upgradedHttpConnection`?

Comment (bartlomieju) [reply ext/http/lib.rs:1089]: Or `httpConnectionUpgrade`?

File: ext/http/websocket_upgrade.rs [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+    // Replace \n with \r\n, but only in headers
+    let (headers, trailing) = s.split_once("\n\n").unwrap();
+    let s = headers.replace("\n", "\r\n") + "\r\n\r\n" + trailing;
+    let s = s.as_ref();
+
+    validate_upgrade_all_at_once(s, expected());
+    validate_upgrade_chunks(s, 1, expected());
+    validate_upgrade_chunks(s, 2, expected());
+    validate_upgrade_chunks(s, 10, expected());
```
Comment (bartlomieju) [inline ext/http/websocket_upgrade.rs:245]: Great tests, thanks!

File: runtime/js/90_deno_ns.js [hunk 7]
```diff
@@ -172,6 +172,7 @@ const denoNsUnstable = {
   funlock: fs.funlock,
   funlockSync: fs.funlockSync,
   upgradeHttp: http.upgradeHttp,
+  upgradeHttp2: http.upgradeHttp2,
```
Comment (bartlomieju) [inline runtime/js/90_deno_ns.js:175]: Let's remove this

**Comment by bartlomieju** [comment]
In https://github.com/denoland/deno/pull/18552 I was able to completely refactor "ext/node" to not rely on Flash server impl and instead use "Deno.serveHttp" with the API provided in this PR. I think we should proceed with it, especially since we don't need to expose it as a public API.

File: ext/http/01_http.js (hunk 2 above)
Comment (mmastrac) [reply ext/http/01_http.js:485]: Excellent idea.

**Review by bartlomieju — APPROVED** [review]
LGTM, exciting change!

**Pull request merged**